src/operators/products.c

```diff
@@ -0,0 +1,221 @@
+/*
+   IGraph library.
+   Copyright (C) 2025  The igraph development team <igraph@igraph.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "igraph_constructors.h"
+#include "igraph_operators.h"
+
+#include "igraph_conversion.h"
+#include "igraph_interface.h"
+
+#include "core/interruption.h"
+#include "math/safe_intop.h"
+
+
+static igraph_error_t cartesian_product(igraph_t *res,
+   const igraph_t *g1, const igraph_t *g2) {
+      
+   igraph_bool_t directed1 = igraph_is_directed(g1);
+   igraph_bool_t directed2 = igraph_is_directed(g2);
+
+   if (directed1 != directed2) {
+      IGRAPH_ERROR("Cartesian product between a directed and an undirected graph is invalid. "
+         "Please ensure that both graphs have the same directionality.", IGRAPH_EINVAL);
+   }
+  
+   igraph_bool_t directed = directed1 && directed2;
+
+   igraph_integer_t vcountg1 = igraph_vcount(g1);
+   igraph_integer_t vcountg2 = igraph_vcount(g2);
+   igraph_integer_t ecountg1 = igraph_ecount(g1);
+   igraph_integer_t ecountg2 = igraph_ecount(g2);
+   igraph_integer_t vcount;
+   igraph_integer_t ecount;
+   igraph_integer_t temp;
+
+   IGRAPH_SAFE_MULT(vcountg1, vcountg2, &vcount);
+   igraph_vector_int_t edges;
+
+   // new edge count = vcountg1*e2 + vcountg2*e1
+   IGRAPH_SAFE_MULT(vcountg1, ecountg2, &ecount);
+   IGRAPH_SAFE_MULT(vcountg2, ecountg1, &temp);
+   IGRAPH_SAFE_ADD(ecount, temp, &ecount);
+   IGRAPH_SAFE_MULT(ecount, 2, &ecount);
+   IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, ecount);
+
+   igraph_integer_t from, to;
+   igraph_integer_t i, j;
+   
+   // Vertex ((i, j)) with i from g1, and j from g2
+   //   will have new vertex id: i * vcountg2 + j
```

Comment:
This should be included in the documentation.

Let's document the vertex ordering for all product types, and keep them consistent with each other.